Handle WORKDIR with dangling target

TomSweeneyRedHat opened the pull request:
When the WORKDIR is a symlink and the target directory does not exist, RUN
commands would fail.  If the target doesn't exist, create the directory in
the container.    This also converts the code to use SecureJoin so that if
a symlink is in the middle of a filepath, it will be converted correctly.

Fixes: #1569

Tip of the hat to @karelyatin for the initial fix

Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

rhatdan commented on imagebuildah/build.go:
```diff
@@ -1817,13 +1817,20 @@ func (b *Executor) deleteSuccessfulIntermediateCtrs() error {
+	// If not, make it.
+	if joinedPath.Mode()&os.ModeSymlink != 0 {
+		if symTarget, err2 := os.Readlink(targetPath); err2 == nil {
+			os.MkdirAll(filepath.Join(s.mountPoint, symTarget), 0755)
```
I believe you have to make sure that the path is withing the s.mountPoint,  Otherwise the link could create a directory on the host.

rhatdan replied:
libpod is using SecureJoin.
./vendor/github.com/cyphar/filepath-securejoin/join.go

TomSweeneyRedHat replied:
@rhatdan thanks for the pointer.  I was going through how to code that last night in my noggin, looks like it's already there with securejoin.

TomSweeneyRedHat commented:
Hold reviews on new patch, just spotted an error, will repush

TomSweeneyRedHat commented:
Repushed a new patch, added SecureJoin and a new test.  All green test buttons.

vrothberg commented on imagebuildah/build.go:
```diff
@@ -480,22 +481,21 @@ func (s *StageExecutor) volumeCacheRestore() error {
+		if err != nil {
+			return errors.Wrapf(err, "error resolving symlinks for copy destination %s", copy.Dest)
+		}
+		targetDest := strings.Split(finalPath, s.mountPoint)
```
I don't think this additional check is necessary but maybe I am missing a corner case. If we keep it, could we use [strings.HasPrefix()](https://golang.org/pkg/strings/#HasPrefix) instead? Using strings.Split() could fail when the mountpoint is equal to a substring of the copy.Dest (e.g., mountpoint=/a/b and copy.Dest=c/a/b/d/a/b).

TomSweeneyRedHat replied:
Just double checking you're commenting on the check at line 494 for the len right?  I need to do the split as finalpath at this point would look something like `/var/lib/docker/overlay2/f92fd1c6882516f93635319c0018706ddb46d3aa85283e94b2012ef6aa25d50d/merged/var/lib/tempest/` and we only would want to pass down `/var/lib/tempest` as code further down the stream will prepend the rest of that path back onto it.
The len check in probably superfluous, but I've been burned a few time without checks like this.

I'll play with HasPrefix and will see if I can get that to go instead, good thought thanks!

giuseppe replied:
to use HasPrefix we need to be sure `s.mountPoint` is also resolved, can we be sure at this point that s.mountPoint has no symlinks and no `..` components?

TomSweeneyRedHat replied:
I'd assumed, possibly badly, that s.mountPoint was completely under our (Buildah's) control at this point and a symlink would not be part of that equation.  @nalind thoughts?  Should I verify that s.mountPoint doesn't have a symlink before the hasPrefix?

TomSweeneyRedHat commented:
I've added a strings.HasPrefix() check before the error and also replaced 'tom' with 'lowerdir' in the tests and related Dockerfiles.  So much for infamy!

TomSweeneyRedHat commented:
Fixes: #1577
also

giuseppe commented on tests/bud.bats:
```diff
@@ -1186,6 +1186,47 @@ load helpers
+  run_buildah --debug=false run ${ctr} ls /tempest
+  expect_output --substring "Dockerfile-2"
+
+  run_buildah --debug=false run ${ctr} ls -alF /etc/notareal.conf 
```
nit: trailing whitespace

TomSweeneyRedHat replied:
fixed

TomSweeneyRedHat commented:
@giuseppe @vrothberg PTAL, I think I've addressed your concerns with the latest.

rhatdan commented:
LGTM

karelyatin commented on imagebuildah/build.go:
```diff
@@ -1816,8 +1818,8 @@ func (b *Executor) deleteSuccessfulIntermediateCtrs() error {
-	targetPath := filepath.Join(s.mountPoint, path)
-	_, err := os.Lstat(targetPath)
+	targetPath, err := securejoin.SecureJoin(s.mountPoint, path)
+	_, err = os.Lstat(targetPath)
```
os.Stat should also work now as securejoin returns resolved path. iirc Lstat was put earlier as targetPath could be a symlink.

TomSweeneyRedHat replied:
Thanks!  Yeah, we probably being too cautious there with the Lstat, but I think I'll keep it as is just in case.

vrothberg commented on imagebuildah/build.go:
```diff
@@ -480,22 +481,23 @@ func (s *StageExecutor) volumeCacheRestore() error {
+			return errors.Wrapf(err, "error resolving symlinks for copy destination %s", copy.Dest)
+		}
+		targetDest := strings.Split(finalPath, secureMountPoint)
+		if !strings.HasPrefix(finalPath, secureMountPoint) || len(targetDest) != 2 {
```
The usage of `strings.Split(...)` is still problematic as the code will fail for reoccurring patterns. Instead, we can use [strings.TrimPrefix](https://golang.org/pkg/strings/#TrimPrefix):
```golang
if !HasPrefix(...) { return error}
copy.Dest = trimPrefix(finalPath, secureMountPoint)
```

TomSweeneyRedHat commented:
Repushed with @vrothberg 's requested HasPrefix/TrimPrefix change.

TomSweeneyRedHat commented:
All happy green test buttons....  @vortherberg, @rhatdan ready to merge?

vrothberg commented:
Sorry, it slipped through my emails.

LGTM, thanks a lot @TomSweeneyRedHat :+1:

rhatdan commented:
@rh-atomic-bot r+
